msi-ec.c

```diff
@@ -367,51 +653,41 @@ static ssize_t shift_mode_show(struct device *device,
 	u8 rdata;
 	int result;
 
-	result = ec_read(MSI_EC_SHIFT_MODE_ADDRESS, &rdata);
+	result = ec_read(conf->shift_mode.address, &rdata);
 	if (result < 0)
 		return result;
 
-	switch (rdata) {
-	case MSI_EC_SHIFT_MODE_TURBO:
-		return sprintf(buf, "%s\n", "turbo");
-	case MSI_EC_SHIFT_MODE_SPORT:
-		return sprintf(buf, "%s\n", "sport");
-	case MSI_EC_SHIFT_MODE_COMFORT:
-		return sprintf(buf, "%s\n", "comfort");
-	case MSI_EC_SHIFT_MODE_ECO:
-		return sprintf(buf, "%s\n", "eco");
-	case MSI_EC_SHIFT_MODE_OFF:
-		return sprintf(buf, "%s\n", "off");
-	default:
+	if (rdata == conf->shift_mode.off_value)
+		return sprintf(buf, "-1\n");
+
+	unsigned int mode = rdata - conf->shift_mode.base_value;
+	if (mode > conf->shift_mode.max_mode)
 		return sprintf(buf, "%s (%i)\n", "unknown", rdata);
-	}
+
+	return sprintf(buf, "%i\n", mode);
```

Comment:
3. This refactor around shift modes downgraded the current behaviour. Also, as user, a plain number isn't informative about the current shift mode. We need debate this point.
a) Do you pretend iterate the shift mode in order to change it without specifying a concrete mode? For example: `echo cycle > /sys/devices/platform/msi-ec/shift_mode` cycles between off -> eco -> comfort -> sport -> turbo
b) Do you like a sysfs path to get all supported modes?  For example: `cat /sys/devices/platform/msi-ec/available_shift_modes`

Reply:
I did this because laptops have different modes for same values (e.g. GF75 uses 0xc0 for sport mode, while Modern 15 uses it for overclock mode).

Now that I think of that, configurations should provide supported modes' names and values and (b) should be implemented. Iterating could be done by userspace apps.

Reply:
> configurations should provide supported modes' names and values and (b) should be implemented

Done

Reply:
```
if (rdata == 0x80)
		return sprintf(buf, "%s\n", "unspecified");
```
Previously, MSI_EC_SHIFT_MODE_OFF = 0x80. I don't remember if shift modes can be off in my laptop, I need install Windows again to verifying it.

In your laptop, what means? Can you reach 0x80 shift mode value using MSI Dragon Center?

Reply:
Look at the comment with graphs below, I explained it there